### ISSUE TITLE
How to reuse Socket instances to avoid the "too many open files" problem?

### ISSUE THREAD

**Issue by andreasweber** [zeromq/jeromq#337]
I want to reuse REQ `ZMQ.Socket` instances to avoid the "too many open files" problem.

I found discussions about socket pools for zermoq, e.g.:
[http://grokbase.com/t/zeromq/zeromq-dev/1274sngtsv/how-to-safely-use-a-socket-in-multiple-threads-in-zmq-2-1-11]

But I'm still not sure what the preconditions in _**jeromq**_ are to share `ZMQ.Socket` instances between threads.
Is it enough to make sure that a Socket is not used by multiple threads at the same time?  

My current scenario in short:
- I have a Singleton class where all the (ZeroMQ) communicaton is initiated, let's call it `ZeroMqSingleton.`
- `ZeroMqSingleton` holds the (one) instance of `ZMQ.Context.` 
- On application shutdown `ZeroMqSingleton.close()` is called and the `ZMQ.Context` is terminated.
- There is a `ZeroMqSingleton.execute(remoteEndpoint)` method which for each call creates a REQ socket, connects it to the endpoint, sends a message, receives a response, closes the REQ socket.
- `ZeroMqSingleton.execute()` is called by threads from a (jetty) thread pool, which is not under my control.
- Under heavy load, the `execute()` calls create/close a lot of REQ sockets in a short period of time, and this causes the "too many open files" problem.

Idea 1:
- Use a REQ pool as described in the link above
- Drawback: Minor performance due to synchronization when getting/releasing REQ pool entries. 

Idea 2: 
- `ZeroMqSingleton` holds a <code>Map<Key, ZMQ.Socket></code>, where "Key" is a string: Thread-Id + remoteEndpoint 
- `execute()` reuses the map entries. It only creates and connects a new REQ socket, if it can't find an entry in the map.
- if a socket.send/recv fails in `execute()` (for whatever reason), the REQ socket is closed and removed from the map.
- On application shutdown, all Socket entries of the map are closed before Context is terminated.
  - Can this be done by the "closing-thread" although the Sockets were created by other threads?
  - (BTW, that's the reason why I can't use ThreadLocal instead of the Map, because then I'd have no chance to close the sockets.)
- Ok, threads from the jetty pool can die, so I have to find a way to avoid resource leaks by cleaning up unused map entries.
- And if a thread died, its Thread-Id could be reused for a new thread. 
  - In that case, can this new thread reuse the REQ socket created by the dead thread?

Do these ideas sound more like patterns or more like antipatterns? Are there better patterns?

**Comment by JasonWu2010** [comment]
Seeing the same issue as all closed tcp connections are in TIME_WAIT state so I guess that sockets have been used up in system.

**Comment by pboettch** [comment]
TLDR; Instead of using `map<thread_id_t, socket>` I use a `static thread_local socket` in C++11.

**Comment by daveyarwood** [comment]
@andreasweber I just closed #596 and it made me think to look for other open issues related to the "Too many open files" exception.

The conclusion I came to in #596 is that getting a "Too many open files" exception is an indication that your architecture needs to be improved such that you're not constantly creating and destroying sockets. To me, the ideas you describe sound like anti-patterns because they encourage one to do exactly that. Instead, sockets should be reused as much as possible, and should be thought of as static rather than disposable.

A pattern I've used successfully is to use the Singleton pattern to declare a `private static Socket socket`, and a `getSocket` method that will create the socket only if it hasn't been done already, otherwise it will reuse the same socket.

I'm still not 100% sure whether or not it is necessary to destroy and recreate the socket for each retry if a response to a request isn't received within a timeframe. The zguide recommends doing that, and I've used that approach successfully; I think the key is to still reuse sockets as much as possible, and only create a new socket as a last resort if the response to a request doesn't come back after several retries. And of course, also keep the interval between retries reasonable; if it's too short, you are more likely to run up against the "Too many open files" problem.

I'm going to close this issue because I don't think the "Too many open files" problem is an issue with JeroMQ specifically, but I'm happy to keep the conversation going re: recommended usage of ZMQ.
